snapcraft/internal/elf.py

```diff
@@ -91,6 +104,65 @@ def __init__(self, *, path: str, magic: str) -> None:
         self.path = path
         self.is_executable = 'interpreter' in magic
         self.dependencies = set()  # type: Set[Library]
+        self.symbols = self._get_symbols(path)  # type: List[Symbol]
+
+    def _get_symbols(self, path) -> List[Symbol]:
+        symbols = list()  # type: List[Symbol]
+        symbols_section = subprocess.check_output([
+            'readelf', '--wide', '--dyn-syms', path]).decode()
+        # Regex inspired by the regex in lintian to match entries similar to
+        # number '1' from the following sample output
+        #
+        # Symbol table '.dynsym' contains 2281 entries:
+        #   Num:    Value          Size Type    Bind   Vis      Ndx Name
+        #     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
+        #     1: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND endgrent@GLIBC_2.2.5 (2)  # noqa
+        symbol_match = (r'\s*\d+:\s*[0-9a-f]+\s*\d+\s*\S+\s*\S+\s*\S+\s*'
+                        r'(?P<section>\S+)\s*(?P<symbol>\S+).*\Z')
+        regex = re.compile(symbol_match)
+        for line in symbols_section.split('\n'):
+            m = regex.search(line)
+            if m:
+                section = m.group('section')
+                try:
+                    name, version = m.group('symbol').split('@')
+                except ValueError:
+                    name = m.group('symbol')
+                    version = ''
+                symbols.append(Symbol(name=name, version=version,
+                                      section=section))
+        return symbols
+
+    def is_linker_compatible(self, *, linker: str) -> bool:
+        """Determines if linker will work given the required glibc version.
+
+        The linker passed needs to be of the format <root>/ld-<X>.<Y>.so.
+        """
+        version_required = self.get_required_glibc()
+        m = re.search(r'ld-(?P<linker_version>[\d.]+).so$',
+                      os.path.basename(linker))
+        if not m:
+            # This is a programmatic error, we don't want to be friendly
+            # about this.
+            raise EnvironmentError('{!r} is incorrect.')
```

Comment:
If you really didn't want to be friendly, you could just use `m` and let a "None has no attribute 'group'" error be thrown by Python itself. Since you're trying to be MORE friendly than that, I suggest a slightly more helpful exception message, perhaps '{!r} is not a valid linker path'? Also, you're probably missing a `.format()` call.

Reply:
nice catch. I want it to be developer friendly, not user friendly :smile: